frontend/utils/image.js

```diff
@@ -28,3 +29,88 @@ export const imageUpload = async (imageFile: File, params: ?Object): Promise<Obj
   const { download } = await sbp('chelonia/fileUpload', imageFile, { type: file.type, cipher: 'aes256gcm' }, params)
   return download
 }
+
+// Image compression
+
+export function supportsWebP (): Promise<boolean> {
+  // Uses a very small webP image to check if the browser supports 'image/webp' format.
+  // (reference: https://developers.google.com/speed/webp/faq#in_your_own_javascript)
+  const verySmallWebP = 'data:image/webp;base64,UklGRiIAAABXRUJQVlA4IBYAAAAwAQCdASoBAAEADsD+JaQAA3AAAAAA'
+  const img = new Image()
+
+  return new Promise(resolve => {
+    img.onload = () => { resolve(img.height > 0) }
+    img.onerror = (e) => { resolve(false) }
+    img.src = verySmallWebP
+  })
+}
+
+function loadImage (url): any {
+  const imgEl = new Image()
+
+  return new Promise((resolve) => {
+    imgEl.onload = () => { resolve(imgEl) }
+    imgEl.src = url
+  })
+}
+
+function generateImageBlobByCanvas ({
+  sourceImage,
+  resizingFactor,
+  quality,
+  compressToType
+}) {
+  const { naturalWidth, naturalHeight } = sourceImage
+  const canvasEl = document.createElement('canvas')
+  const c = canvasEl.getContext('2d')
+
+  canvasEl.width = naturalWidth * resizingFactor
+  canvasEl.height = naturalHeight * resizingFactor
+
+  c.drawImage(
+    sourceImage,
+    0,
+    0,
+    canvasEl.width,
+    canvasEl.height
+  )
+
+  return new Promise((resolve) => {
+    canvasEl.toBlob(blob => {
+      resolve(blob)
+    }, compressToType, quality)
+  })
+}
+
+export async function compressImage (imgUrl: string, sourceMimeType?: string): Promise<any> {
+  // Takes a source image url and generate a blob of the compressed image.
+
+  // According to the testing result, 0.8 is a good starting point for both resizingFactor and quality for .jpeg and .webp.
+  // For other image types, we use 0.9 as the starting point.
+  const defaultFactor = ['image/jpeg', 'image/webp'].includes(sourceMimeType) ? 0.8 : 0.9
+  let resizingFactor = defaultFactor
+  let quality = defaultFactor
+  // According to the testing result, webP format has a better compression ratio than jpeg.
+  const compressToType = await supportsWebP() ? 'image/webp' : 'image/jpeg'
+  const sourceImage = await loadImage(imgUrl)
+
+  while (true) {
+    const blob = await generateImageBlobByCanvas({
+      sourceImage,
+      resizingFactor,
+      quality,
+      compressToType
+    })
+    const sizeDiff = blob.size - IMAGE_ATTACHMENT_MAX_SIZE
+
+    if (sizeDiff <= 0 || // if the compressed image is already smaller than the max size, return the compressed image.
+      quality <= 0.3) { // Do not sacrifice the image quality too much.
```

Comment:
From the multiple tests I've done, when the quality factor value passed to `canvas.toBlob()` becomes `<= 0.3`, the outcome image starts to lose the quality to an unpleasant level. `0.3` here is a value I chose subjectively, but feel free to try compression test yourself using [this codepen](https://codepen.io/freenaturalsoul/pen/vYoMWWv) I created for this task and let me know if you would like a change.